galley/pkg/config/analysis/analyzers/auth/mtls.go

```diff
@@ -0,0 +1,333 @@
+		return true
+	})
+
+	c.ForEach(metadata.IstioAuthenticationV1Alpha1Policies, func(r *resource.Entry) bool {
+		ns, _ := r.Metadata.Name.InterpretAsNamespaceAndName()
+		namespaces[ns] = struct{}{}
+
+		err := pc.AddPolicy(r, r.Item.(*v1alpha1.Policy))
+		if err != nil {
+			// AddPolicy can return a NamedPortInPolicyNotFoundError - if it
+			// does we can print a useful message.
+			// TODO this should be in its own analyzer, and ignored here.
+			if missingPortNameErr, ok := err.(mtls.NamedPortInPolicyNotFoundError); ok {
+				c.Report(metadata.IstioAuthenticationV1Alpha1Meshpolicies,
+					msg.NewPolicySpecifiesPortNameThatDoesntExist(r, missingPortNameErr.PortName, missingPortNameErr.FQDN))
+				return true
+			} else {
+				c.Report(metadata.IstioAuthenticationV1Alpha1Meshpolicies, msg.NewInternalError(r, err.Error()))
+				return false
+			}
+		}
+		return true
+	})
+
+	drc := mtls.NewDestinationRuleChecker()
+	c.ForEach(metadata.IstioNetworkingV1Alpha3Destinationrules, func(r *resource.Entry) bool {
+		ns, _ := r.Metadata.Name.InterpretAsNamespaceAndName()
+		namespaces[ns] = struct{}{}
+
+		drc.AddDestinationRule(r, r.Item.(*v1alpha3.DestinationRule))
+		return true
+	})
+
+	// Here we explicitly handle the common case where a user specifies a
+	// MeshPolicy with no global DestinationRule. We also track if we report a
+	// problem with the global configuration. This is used later to suppress
+	// reporting a message for every service/namespace combination due to the
+	// same misconfiguration.
+	globalMTLSMisconfigured := false
+	mpr := pc.MeshPolicy()
+	globalMtls, globalDR := drc.DoesNamespaceUseMTLSToService("istio-system", "istio-system", mtls.NewTargetService("*.svc.cluster.local"))
+	if mpr.MTLSMode == mtls.ModeStrict && !globalMtls {
+		// We may or may not have a matching DR. If we don't, use
+		// the special string (none)
+		globalDRName := "(none)"
+		if globalDR != nil {
+			globalDRName = globalDR.Origin.FriendlyName()
+		}
+		c.Report(
+			metadata.IstioAuthenticationV1Alpha1Meshpolicies,
+			msg.NewMTLSPolicyConflict(
+				mpr.Resource,
+				"*.svc.cluster.local",
+				"istio-system",
+				globalDRName,
+				globalMtls,
+				mpr.Resource.Origin.FriendlyName(),
+				mpr.MTLSMode.String()))
+		globalMTLSMisconfigured = true
+	}
+
+	// Also handle the less-common case where a global DR exists that specifies
+	// mtls, but MTLS is off
+	if mpr.MTLSMode == mtls.ModePlaintext && globalMtls {
+		// We may or may not have a matching policy. If we don't, use the
+		// special string (none)
+		globalPolicyName := "(none)"
+		if mpr.Resource != nil {
+			globalPolicyName = mpr.Resource.Origin.FriendlyName()
+		}
+		c.Report(
+			metadata.IstioNetworkingV1Alpha3Destinationrules,
+			msg.NewMTLSPolicyConflict(
+				globalDR,
+				"*.svc.cluster.local",
+				"istio-system",
+				globalDR.Origin.FriendlyName(),
+				globalMtls,
+				globalPolicyName,
+				mpr.MTLSMode.String()))
+		globalMTLSMisconfigured = true
+	}
```

Comment:
I considered it, but I think it would be worse if I did. I don't think I can name the helper function well (which is usually a sign that its a bad abstraction), and it depends on an awful lot of context to build the analyzer message.